### PR TITLE
check if firewall is connected to VPN

### PR DESCRIPTION
**Pull request by GrigoriyMikhalkin**

### CODE REVIEW

File: cmd/metal-api/internal/headscale/client.go [hunk 1]
```diff
@@ -119,6 +119,29 @@ func (h *HeadscaleClient) CreatePreAuthKey(namespace string, expiration time.Tim
 	return resp.PreAuthKey.Key, nil
 }
 
+func (h *HeadscaleClient) DescribeMachine(hostname, projectID string) (connected bool, err error) {
```
Comment (majst01) [inline cmd/metal-api/internal/headscale/client.go:122]: LGTM, the only downside actually is that hostname is not guaranteed to be unique in a project. It is the case for our clusters but in the pure metal use case it is not. Should be at least documented.

Comment (majst01) [reply cmd/metal-api/internal/headscale/client.go:122]: One idea would be to add a tag from the firewall during connecting with:

`tailscale up --advertise-tags tag:metal-stack.io/machineid=<id of the firewall> `

and then check if the machineid tag with exists

Comment (majst01) [reply cmd/metal-api/internal/headscale/client.go:122]: Yes sure

Comment (GrigoriyMikhalkin) [reply cmd/metal-api/internal/headscale/client.go:122]: Situation is even more complicated then i initially thought. Headscale API refuses to return any tags if ACL policy isn't set. So tags are very tightly connected to ACLs.

File: cmd/metal-api/internal/headscale/client.go [hunk 2]
```diff
@@ -119,6 +119,30 @@ func (h *HeadscaleClient) CreatePreAuthKey(namespace string, expiration time.Tim
...
+	req := &headscalev1.ListMachinesRequest{
+		Namespace: projectID,
+	}
+	resp, err := h.client.ListMachines(h.ctx, req)
+	if err != nil || resp == nil {
+		return false, fmt.Errorf("failed to list machines: %w", err)
+	}
+
+	for _, m := range resp.Machines {
+		if m.Name == machineid {
```
Comment (majst01) [inline cmd/metal-api/internal/headscale/client.go:132]: i thought a check for the tag containing the machineid is now possible, why is the name now equal to machineid ?

Comment (GrigoriyMikhalkin) [reply cmd/metal-api/internal/headscale/client.go:132]: https://github.com/metal-stack/metal-api/pull/336#discussion_r975745133, unfortunately it isn't possible if ACL policy isn't set.

**Review by majst01 — APPROVED** [review]

**Pull request merged**